src/v/redpanda/application.cc

```diff
@@ -3113,6 +3114,25 @@ void application::start_runtime_services(
           .get();
     }
 
+    // Verify the enterprise license when trying to upgrade Redpanda.
+    // By this point during startup we have enough information to evaluate both
+    // the state of the license and what enterprise features are used.
+    // - If redpanda has been restarted on an existing node, we have already
+    //   loaded the feature table from the local snapshot in
+    //   application::load_feature_table_snapshot and replayed the local
+    //   controller log in controller::start.
+    // - If this is a new node joining an existing cluster, by this point we
+    //   have received a controller snapshot from another node in the join
+    //   response and have replicated and replayed the the controller stm to the
+    //   last_applied offset received in the join_node_reply above.
+    controller->get_feature_manager()
+      .invoke_on(
+        cluster::feature_manager::backend_shard,
+        [](cluster::feature_manager& fm) {
+            return fm.verify_enterprise_license();
+        })
+      .get();
```

Comment:
Presumably the proper we want is that start-up happens in phases.

* Phase 1 - Startup to the point that "active version" is set and correct and pause.
* Phase 2 - Check license, and abort start-up if it is not valid
* Phase 3 - Allow the feature manager the opportunity agree on a newer version for the cluster

Is this what is happening? If so, great, but it's kind of hard to "see" that this happening. The phases are spread out, and maybe even some of them are asynchronously running?

 I guess the first thing to do is make sure we're on the same page about what is happening. Then we can figure out if there is any changes needed. The minor concern I have is that it will be easy to break this with even simple code refactoring?

Reply:
> Is this what is happening?

Yes.

I think the complexity comes from:
1. The fact that Phase 1 can take separate paths depending on whether this is a restarting node or a node that has just joined the cluster. I've tried to describe this in the comment above.
2. In Phase 3, while the application startup itself is sequential, the fibre that performs the potential advancing of a newer version for the cluster runs asynchronously, so we need to signal to that fibre which we do with a mutex.

> The minor concern I have is that it will be easy to break this with even simple code refactoring?
Do you have a specific example in mind? The main behaviour we care about is that non-compliant clusters abort and compliant clusters don't abort during an upgrading start, which I've tried to provide reasonable coverage for in the new ducktape tests.